acpi: Multiple cleanups, fixes and clarifications

jhedberg opened the pull request:
There were several inconsistencies, unnecessary code, as well as clearly incorrect code in acpi.c. This PR tries to address some low hanging fruits - there's likely more to come. @najumon1980 sorry for the conflicts this will cause with your PR, but it seems like that one will take longer to go into the tree.

Commit:
acpi: Remove unnecessary static array

The acpi_tables array is only needed for systems where dynamic memory
allocation is not available during the early ACPI init phase. In the
Zephyr case we can immediately start using k_malloc, so this is
unnecessary.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
acpi: Remove unused Kconfig variable

ACPI currently uses implicit (auto) initialization, i.e. it doesn't need
any init level or priority.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
acpi: Don't include PCI PRT code if not enabled in Kconfig

If the CONFIG_PCIE_PRT option is disabled it makes no sense to bloat the
ACPI build with PRT-related code or static tables. The diff looks a bit
larger since functions in acpi.c had to be shuffled around to be able to
be included in a single "#ifdef CONFIG_PCIE_PRT" block.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
acpi: Rename bus_ctx to simply acpi

ACPI is not treated as a bus in Zephyr. Rename the global context from
bus_ctx to simply acpi.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
acpi: Fix status variable type

Use the appropriate ACPI_STATUS type for any status variable that stores
return values from ACPICA APIS.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
acpi: Fix acpi_get_irq_routing_table() rt_size parameter meaning

The acpi_get_irq_routing_table() takes a pointer to an array of
ACPI_PCI_ROUTING_TABLE elements rather than a generic buffer pointer
(e.g. void *).

Because of the above, it makes sense to specify the array size as an
actual ARRAY_SIZE() value, since it makes no sense to accept buffers
which are not a multiple of sizeof(ACPI_PCI_ROUTING_TABLE) long.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
acpi: Move local pci_prt_table processing to the right place

The acpi_get_irq_table() function takes a pointer to a table that can
come from anywhere, i.e. it doesn't have to be the acpi.pci_prt.table
that acpi.c uses. Because of this, the correct place to iterate and
process the acpi.pci_prt_table is in the function that actually passes
acpi.pci_prt_table to the acpi_get_irq_table() function.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
acpi: Don't mix ACPI_STATUS and "int" return types

ACPI_STATUS variables should not store values of any other error domain
(like negative POSIX error codes used for Zephyr's ACPI API).

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

finikorg commented on lib/acpi/acpi_shell.c:
```diff
@@ -267,8 +272,10 @@ SHELL_STATIC_SUBCMD_SET_CREATE(
 	SHELL_CMD(prs, NULL,
 		  "display device possible resource settings (eg:acpi crs _SB.PC00.LPCB.RTC)",
 		  dump_dev_prs),
+#ifdef CONFIG_PCIE_PRT
```
I wonder can we use `SHELL_COND_CMD_ARG`

jhedberg replied:
> I wonder can we use `SHELL_COND_CMD_ARG`

Yes, that sounds like a good idea

jhedberg replied:
> > I wonder can we use `SHELL_COND_CMD_ARG`
> 
> Yes, that sounds like a good idea

Done

jhedberg replied:
> > > I wonder can we use `SHELL_COND_CMD_ARG`
> > 
> > 
> > Yes, that sounds like a good idea
> 
> Done

@finikorg this caused build failures, since `SHELL_COND_CMD()` still requires the the callback to defined (not behind ifdef) and therefore any code in the callback to be buildable. I fixed this by moving the `pci_prt_table` array into the callback, and by putting most of the callback implementation behind `IF_ENABLED(...`. CI should hopefully pass now.

finikorg replied:
OK, noticed myself that that CMD_COND requires that workaround

finikorg commented on lib/acpi/acpi.c:
```diff
@@ -250,12 +247,12 @@ static ACPI_STATUS dev_resource_enum_callback(ACPI_HANDLE obj_handle, UINT32 lev
 
 	AcpiWalkNamespace(ACPI_TYPE_DEVICE, ACPI_ROOT_OBJECT, ACPI_UINT32_MAX,
-			  dev_resource_enum_callback, NULL, bus, NULL);
+			  dev_resource_enum_callback, NULL, NULL, NULL);
```
are just renaming "bus_ctx to simply acpi" here?

jhedberg replied:
The parameter is pa

> are just renaming "bus_ctx to simply acpi" here?

The parameter is just passes as-is to the callback. Since it's the same global variable (internal to acpi.c) the callback doesn't need it. IIRC I did update the callback implementation to not rely on the parameter anymore.

jhedberg replied:
> > are just renaming "bus_ctx to simply acpi" here?
> 
> The parameter is just passed as-is to the callback. Since it's the same global variable (internal to acpi.c) the callback doesn't need to get it as a function "user data" parameter. IIRC I did update the callback implementation to not rely on the parameter anymore.

jhedberg replied:
> are just renaming "bus_ctx to simply acpi" here?

One more clarification: it's not just a rename of the variable - we loose the name for the struct type as well, so it's not just that we don't want to pass around its pointer but that we can't.

finikorg commented on lib/acpi/acpi.c:
```diff
@@ -289,14 +289,13 @@ int acpi_current_resource_get(char *dev_name, ACPI_RESOURCE **res)
 
 	status = check_init_status();
 	if (status) {
-		return status;
```
the check_init_status() itself can return mixed error code

jhedberg replied:
> the check_init_status() itself can return mixed error code

Good catch. Will fix.

jhedberg replied:
> > the check_init_status() itself can return mixed error code
> 
> Good catch. Will fix.

Fixed

finikorg reviewed (commented):
some comments

tbursztyka commented:
finally, getting rid of "bus" name in acpi context!

tbursztyka reviewed (dismissed):
nice

jhedberg commented:
For the record, the CI failure is unrelated to this PR, most likely due to #64544 :
```
arduino_uno_r4_minima     samples/synchronization/sample.kernel.synchronization  ERROR
```
I'll trigger a rerun once a fix has been merged.

Commit:
acpi: shell: Remove unnecessary newlines

The ACPI shell command create way too much vertical empty space. Remove
the unnecesary newlines and use indentation to indicate grouping of
lines. At the same time, place case statments with variable declarations
behind {} since otherwise both the Zephyr compliance checker and some
other static analyzers get confused by the code in the branch.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
@finikorg @tbursztyka CI is all green now. Please re-approve :)

Commit:
acpi: Remove unnecessary prs_buffer variable

This was never used for anything, since the ACPI API overwrites the
pointer when fetching a resource list.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
acpi: Remove redundant newline characters from logs

The log functions themselves automatically add newline characters, so no
need to do it when calling the log macros.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
acpi: Use ACPICA typedef names when available

For most types, ACPICA provides both a struct name as well as a typedef.
The struct names follow the exact same naming style as Zephyr's ACPI
API, which makes it impossible to distinguish which type is defined by
Zephyr and which comes from ACPICA. It's therefore better to use the
typedefs, since they follow a distinct style compared to the Zephyr API.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
Just pushed a few more cleanups that I had piled up locally, related to excessive newlines in shell and log output, one redundant variable removal, as well as switch to ACPICA typdefs to make it clear where each type is coming from (ACPICA or Zephyr ACPI)

Commit:
acpi: shell: Fix specifying string precision

The ACPI table signature is not null terminated, so a precision needs to
be provided in format strings. There was an attempt to do this, but it
was done in an incorrect way, which resulted in garbage characters
getting printed.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
arch: x86: Use ACPICA typdef instead of struct name

Use ACPI_MADT_LOCAL_APIC instead of struct acpi_madt_local_apic. In the
same go, switch to IF_ENABLED from ifdef - slightly more readable, and
this keeps some static analyzers happy (e.g. upstream Compliance check).

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

finikorg commented on lib/acpi/acpi.c:
```diff
@@ -203,7 +203,7 @@ static ACPI_STATUS dev_resource_enum_callback(ACPI_HANDLE obj_handle, UINT32 lev
 	ACPI_DEVICE_INFO *dev_info;
 
-	LOG_DBG("%s %p\n", __func__, node);
+	LOG_DBG("%s %p", __func__, node);
```
Do we need `__func__` ?

jhedberg replied:
> Do we need `__func__` ?

Probably not, as that can be enabled in the log subsystem too. I'll leave it as an exercise to be fixed in a separate PR however (already enough commits in this one).

najumon1980 commented:
Sure, I can rebase once this PR get merged.

najumon1980 commented on arch/x86/core/intel64/cpu.c:
```diff
@@ -144,14 +144,14 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
-		x86_cpu_loapics[cpu_num] = lapic->Id;
-	}
-#endif
+	IF_ENABLED(CONFIG_ACPI, ({
```
If ACPICA module not included in west this can fail ?

najumon1980 replied:
also #include <acpi.h> is under CONFIG_ACPI

jhedberg replied:
> If ACPICA module not included in west this can fail ?

No, it will still work (as you can also see from CI passing). There is also no problem with the #ifdef guard for acpi.h. IF_ENABLED works the same way as #ifdef, i.e. it's not the same as IS_ENABLED.

For the future, I'm thinking that maybe we should split Zephyr's acpi.h into two separate headers: a public and an internal one. The public one would not depend on any ACPI definitions, so this would be the header that any c-file would use that's not 100% ACPI related (like this cpu.c). We would however need to create some new wrappers/abstractions to allow access to all necessary functionality this way.
